src/kirin/rewrite/infer_len.py

```diff
@@ -0,0 +1,41 @@
+from functools import cached_property
+
+from kirin import ir, types
+from kirin.analysis import const
+from kirin.dialects import py
+from kirin.rewrite.abc import RewriteRule, RewriteResult
+
+
+class InferLen(RewriteRule):
+
+    @cached_property
+    def IListType(self):
+        from kirin.dialects import ilist
+
+        return ilist.IListType
```

Comment:
I don't understand what is this for? Also shouldn't this rewrite pass be inside `ilist`? It does not necessarily always happen during an aggressive fold.

Reply:
The problem is circular imports and it seems bad to have to import this type every time a method is called.

Reply:
If it is not needed for const folding then we don't need this rewrite. 

The only case I can think of if you are aggressively folding a kernel with an argument that has a type annotation that exactly bounds the size of that input. In that case you can unroll the kernel if you're loop bounds depend on the size of the argument not necessarily the argument itself. 

it is a pretty small edge case so sure we can move it into ilist to prevent the circular import problems.